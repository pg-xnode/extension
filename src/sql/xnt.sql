set search_path to xml;

CREATE TABLE templates (
	id integer NOT NULL,
	template xnt
);

INSERT INTO templates VALUES 
-- the simplest case: only attribute values are parametrized
(0, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{$p1}" y="{$p2}"/></xnt:template>'),
-- in addition, parameter is used to generate the element name as well as attribute names
-- this template also shows that attributes can have different namespace prefixes
-- as long as they point to the same URI
(1, '<xnt:template
	xmlns:xnt="http://www.pg-xnode.org/xnt"
	xmlns:xnt_2="http://www.pg-xnode.org/xnt">
<xnt:element name="point_{$p1}">
	<xnt:attribute xnt:name="x_{$p1}" xnt_2:value="{$p2}"/>
	<xnt:attribute name="y_{$p2}" value="{$p3}"/>
</xnt:element>
</xnt:template>'),
-- node can also be used as parameter
(2, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt">
<area><xnt:copy-of expr="$p1"/></area>
</xnt:template>'),
(3, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{1}" y="{2}"/></xnt:template>');

SELECT id, template
FROM templates
ORDER BY id;

CREATE TABLE points (
	id integer NOT NULL,
	x float NOT NULL,
	y float NOT NULL
);

INSERT INTO points VALUES
(0, 0.0, 0.0), (1, 0.3, 5.4), (2, 3.15, 0.01);

CREATE TABLE point_nodes (
	id integer NOT NULL,
	p node NOT NULL
);

INSERT INTO point_nodes
SELECT p.id, xml.node(t.template, '{p1, p2}', ROW (x, y))
FROM templates t, points p
WHERE t.id=0;

SELECT *
FROM point_nodes
ORDER BY id;

SELECT xml.node(template, '{p1, p2, p3}', ROW (p.id, p.x, p.y))
FROM templates t, points p
WHERE t.id=1
ORDER BY p.id;

WITH fragments(f) AS (
	-- order the rows so that the fragment is never different from the expected test output.  
	SELECT xml.fragment(p ORDER BY id)
	FROM point_nodes
)
SELECT xml.node(t.template, '{p1}', ROW(f))
FROM fragments f,
templates t
WHERE t.id=2;

-- Is unary minus processed correctly when used with expression parameters?
select xml.node('<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><a i="{-$x}"/></xnt:template>', '{ x }', ROW( 3.14::float ));
select xml.node('<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><a i="{$x}"/></xnt:template>', '{ x }', ROW( -3.14::float ));
select xml.node('<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><a i="{-$x}"/></xnt:template>', '{ x }', ROW( -3.14::float ));

DROP TABLE templates;
DROP TABLE points;
DROP TABLE point_nodes;
