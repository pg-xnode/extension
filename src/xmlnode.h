/*
 * Copyright (C) 2012, Antonin Houska
 */

#ifndef XMLNODE_H
#define XMLNODE_H

#include <ctype.h>
#include "postgres.h"
#include "fmgr.h"
#include "utils/elog.h"

#define XNODE_CHAR_SPACE		0x20
#define XNODE_CHAR_EXCLMARK		0x21
#define XNODE_CHAR_QUOTMARK		0x22
#define XNODE_CHAR_HASH			0x23
#define XNODE_CHAR_DOLLAR		0x24
#define XNODE_CHAR_PCT			0x25
#define XNODE_CHAR_AMPERSAND	0x26
#define XNODE_CHAR_APOSTR		0x27
#define XNODE_CHAR_LBRKT_RND	0x28
#define XNODE_CHAR_RBRKT_RND	0x29
#define XNODE_CHAR_ASTERISK		0x2a
#define XNODE_CHAR_PLUS			0x2b
#define XNODE_CHAR_COMMA		0x2c
#define XNODE_CHAR_DASH			0x2d
#define XNODE_CHAR_DOT			0x2e
#define XNODE_CHAR_SLASH		0x2f
#define XNODE_CHAR_COLON		0x3a
#define XNODE_CHAR_SEMICOLON	0x3b
#define XNODE_CHAR_LARROW		0x3c
#define XNODE_CHAR_EQ			0x3d
#define XNODE_CHAR_RARROW		0x3e
#define XNODE_CHAR_QUESTMARK	0x3f
#define XNODE_CHAR_AT			0x40
#define XNODE_CHAR_LBRACKET		0x5b
#define XNODE_CHAR_RBRACKET		0x5d
#define XNODE_CHAR_UNDERSCORE	0x5f
#define XNODE_CHAR_LBRKT_CUR	0x7b
#define XNODE_CHAR_PIPE			0x7c
#define XNODE_CHAR_RBRKT_CUR	0x7d

/* How many bytes need to be preserved to achieve given alignment. */
#define MAX_PADDING(alignment)	((alignment) - 1)

#define XNODE_NAMESPACE_DEF_PREFIX		"xmlns"

typedef enum XMLNodeKind
{
	XMLNODE_DOC = 0,
	XMLNODE_DTD,
	XMLNODE_ELEMENT,
	XMLNODE_ATTRIBUTE,
	XMLNODE_COMMENT,
	XMLNODE_CDATA,
	XMLNODE_PI,
	XMLNODE_TEXT,

	/*
	 * Document fragment containing zero or one child node should never be
	 * created. It's relied on that a non-compound node will be used in such
	 * cases.
	 *
	 * Also it's not allowed to create nested fragments (i.e. one fragment to
	 * be a child of another one).
	 */
	XMLNODE_DOC_FRAGMENT,

	XMLNODE_NODE,

	XNTNODE_ROOT,
	XNTNODE_TEMPLATE,
	XNTNODE_COPY_OF,
	XNTNODE_ELEMENT,
	XNTNODE_ATTRIBUTE
} XMLNodeKind;

typedef uint32 XMLNodeOffset;

#define XNODE_ALIGNOF_NODE_OFFSET ALIGNOF_INT

/*
 * Zero is only invalid if interpreted as *relative* offset.
 */
#define XMLNodeOffsetInvalid	0

typedef struct XMLNodeCommonData
{
	/*
	 * XMLNodeKind would be expected here, but char is used to save space.
	 */
	uint8		kind;
	uint8		flags;
} XMLNodeCommonData;

/*
 * For simple (non-compound) node types, content follows the structure immediately
 * (there are no references to children).
 */
typedef XMLNodeCommonData XMLNodeHdrData;

typedef XMLNodeHdrData *XMLNodeHdr;

/*
 * A compound node, i.e node containing other node(s).
 * Used for the following node types:
 * XMLNODE_ELEMENT, XMLNODE_DOC, XMLNODE_DOC_FRAGMENT
 *
 * The header is followed by array of references.
 * Reference means relative offset of a nested element, from the current element backwards (child
 * is always stored at lower addresses than parent).
 * Such a reference takes 1 to 4 bytes, while this 'byte width' must be equal for all references
 * that given compound node keeps. Thus the byte width is derived from the maximum offset the compound
 * node has (given the 'child first' ordering, the maximum offset is that of the first child).
 *
 * XNODE_GET_REF_BWIDTH() / XNODE_SET_REF_BWIDTH() should be used to set the byte width. Direct access
 * to the storage is discouraged.
 *
 * The array of references is followed either by element name (XMLNODE_ELEMENT) or by document header
 * (XMLNODE_DOC).
 */
typedef struct XMLCompNodeHdrData
{
	XMLNodeCommonData common;

	uint16		children;
} XMLCompNodeHdrData;

typedef XMLCompNodeHdrData *XMLCompNodeHdr;

#define XNODE_ALIGNOF_COMPNODE	ALIGNOF_SHORT


typedef struct XMLDeclData
{
	uint8		flags;
	uint8		version;
	uint8		enc;
	bool		standalone;
}	XMLDeclData;

typedef struct XMLDeclData *XMLDecl;

#define XMLDECL_HAS_ENC		(1 << 0)
#define XMLDECL_HAS_SD_DECL (1 << 1)
/*
 * Bits 2, 3 and 4 are reserved for quotation mark type
 */
#define XMLDECL_GET_QUOT_MARK(decl, i) (decl->flags & (1 << (i + 2))) ? XNODE_CHAR_APOSTR :\
		XNODE_CHAR_QUOTMARK;

#define XMLNODE_MAX_CHILDREN	0xFFFF

#define XNODE_WHITESPACE(c) (*(c) == 0x9 || *(c) == 0xA || *(c) == 0xD || *(c) == 0x20)
#define XNODE_VALID_NAME_START(c)	(*(c) == XNODE_CHAR_COLON || *(c) == XNODE_CHAR_UNDERSCORE || \
		isXMLCharInInterval(c, nameStartCharIntervals, XNODE_NAME_START_CHAR_INTERVALS))
#define XNODE_VALID_NAME_CHAR(c)	(XNODE_VALID_NAME_START(c) || *(c) == XNODE_CHAR_DASH ||\
		*(c) == XNODE_CHAR_DOT ||	isXMLCharInInterval(c, nameCharIntervals, XNODE_NAME_CHAR_INTERVALS))
#define XNODE_VALID_CHAR(c) (XNODE_WHITESPACE(c) || isXMLCharInInterval(c, charIntervals, CHAR_INTERVALS))

#define XNODE_NAME_START_CHAR_INTERVALS 15
#define XNODE_NAME_CHAR_INTERVALS		4

#define XNODE_OFFSET(node, doc) ((XMLNodeOffset) ((char *) (node) - (char *) VARDATA(doc)))

extern char getXMLNodeOffsetByteWidth(XMLNodeOffset o);

#define XNODE_FIRST_REF(cnd) ((char *) (cnd) + sizeof(XMLCompNodeHdrData))
#define XNODE_LAST_REF(cnd) (XNODE_FIRST_REF(cnd) + ((cnd)->children - 1) * XNODE_GET_REF_BWIDTH(cnd))

/*
 * Only use this for simple nodes
 */
#define XNODE_CONTENT(nd) ((char *) ((XMLNodeHdr) (nd) + 1))

/*
 * TODO Check if the multiplication needs to be performed in alternative
 * (more efficient) way. The same for XNODE_LAST_REF() above
 */
/*
 * Returns name of XML element, but can also be used to get the appropriate data
 * out of other kinds of compound nodes.
 */
#define XNODE_ELEMENT_NAME(el) (XNODE_FIRST_REF(el) + (el)->children * XNODE_GET_REF_BWIDTH(el))

#define XNODE_NEXT_REF(ptr, cnd) (ptr + XNODE_GET_REF_BWIDTH(cnd))
#define XNODE_PREV_REF(ptr, cnd) (ptr - XNODE_GET_REF_BWIDTH(cnd))
#define XNODE_HAS_CHILDREN(cnd) (cnd->children > 0)

typedef struct varlena xmlnodetype;
typedef xmlnodetype *xmlnode;

extern Datum xmlnode_in(PG_FUNCTION_ARGS);
extern Datum xmlnode_out(PG_FUNCTION_ARGS);
extern Datum xmlnode_kind(PG_FUNCTION_ARGS);
extern Datum xmlnode_debug_print(PG_FUNCTION_ARGS);

extern char *dumpXMLNode(char *data, XMLNodeOffset rootNdOff);

typedef struct varlena xmldoctype;
typedef xmldoctype *xmldoc;

extern Datum xmldoc_in(PG_FUNCTION_ARGS);
extern Datum xmldoc_out(PG_FUNCTION_ARGS);

extern Datum xmlnode_to_xmldoc(PG_FUNCTION_ARGS);
extern Datum xmldoc_to_xmlnode(PG_FUNCTION_ARGS);


/* Get a pointer to root node offset from varlena (xmlnode, xmldoc) value. */
#define XNODE_ROOT_OFFSET_PTR(raw)	((XMLNodeOffset *) ((char *) raw + VARSIZE(raw) - sizeof(XMLNodeOffset)))
/* ... and the root offset itself */
#define XNODE_ROOT_OFFSET(raw) (*XNODE_ROOT_OFFSET_PTR(raw))

/*
 * Get root node from varlena (xmlnode, xmldoc) value
 */
#define XNODE_ROOT(raw) ((XMLNodeHdr) ((char *) VARDATA(raw) + XNODE_ROOT_OFFSET(raw)))

extern XMLNodeOffset readXMLNodeOffset(char **input, unsigned char bytes, bool step);
extern void writeXMLNodeOffset(XMLNodeOffset ref, char **output, unsigned char bytes, bool step);


#define XNODE_CONTAINER_CHUNK	16

typedef enum XNodeListItemKind
{
	XNODE_LIST_ITEM_BOOLEAN,
	XNODE_LIST_ITEM_SINGLE_OFF,
	XNODE_LIST_ITEM_SINGLE_PTR,
	XNODE_LIST_ITEM_RANGE
} XNodeListItemKind;

typedef struct XNodeOffsetRange
{
	XMLNodeOffset lower;
	XMLNodeOffset upper;
} XNodeOffsetRange;

typedef struct XNodeListItem
{
	bool		valid;			/* Only applicable when the containing list is
								 * used as 'ignore list'. */
	XNodeListItemKind kind;

	union
	{
		bool		boolean;
		XMLNodeOffset singleOff;
		void	   *singlePtr;
		XNodeOffsetRange range;
	}			value;
} XNodeListItem;

/*
 * A container to be used as a stack in most cases, but sometimes we used it
 * as FIFO.
 */

typedef struct XMLNodeContainerData
{
	unsigned int size;
	unsigned int position;
	XNodeListItem *content;
} XMLNodeContainerData;

typedef struct XMLNodeContainerData *XMLNodeContainer;


#define UTF_MAX_WIDTH		4

typedef struct UTF8Interval
{
	char		first[UTF_MAX_WIDTH];
	char		last[UTF_MAX_WIDTH];
} UTF8Interval;


extern UTF8Interval nameStartCharIntervals[XNODE_NAME_START_CHAR_INTERVALS];
extern UTF8Interval nameCharIntervals[XNODE_NAME_CHAR_INTERVALS];
extern bool isXMLCharInInterval(char *c, UTF8Interval *intervals, unsigned short int intCount);


/*
 * Flags that are supposed to be generic (i.e. used for 2 or more nodes)
 * must be at the highest positions.
 *
 * Kind-specific flags are expected at the lowest positions.
 */

/*
 * For text node set if it contains predefined reference(s)? (&amp; &lt;
 * ...). For CDATA node set if it contains	'<', '>' or '&'
 */
#define XNODE_TEXT_SPEC_CHARS		(1 << 7)
/* Namespace declaration or use. */
#define XNODE_NMSP_PREFIX			(1 << 6)

/* Bits 0 and 1 indicate maximum byte width of the distance between parent and child */
#define XNODE_REF_BWIDTH				0x03
#define XNODE_EMPTY						(1 << 2)
#define XNODE_DOC_XMLDECL				(1 << 3)

/*
 * Set if the compound element contains something else than (NULL-terminated) name.
 * Some attributes of such special (sometimes called reserved) nodes also usually
 * differ from the name-value pair of NULL-terminated strings.
 */
#define XNODE_EL_SPECIAL				(1 << 4)

#define XNODE_ATTR_APOSTROPHE			(1 << 0)

/* Set if apostrophe is used as value delimiter. */
#define XNODE_ATTR_CONTAINS_REF			(1 << 1)

/*
 * This flag is set when value is stored as a text, but in fact represents a
 * number
 */
#define XNODE_ATTR_NUMBER				(1 << 2)

/* Set if attribute value is not a simple NULL-terminated string. */
#define XNODE_ATTR_VALUE_BINARY			(1 << 3)



#define XNODE_PI_HAS_VALUE			(1 << 0)

#define XNODE_ATTR_IS_NUMBER(node)			(((node)->flags & XNODE_ATTR_NUMBER) != 0)

/*
 * Macros to read / write byte width of the maximum child's offset.
 * The value we store is 0-based, i.e. 0 in the 'flags' filed means 1 byte per reference.
 */
#define XNODE_GET_REF_BWIDTH(cnd)		(((cnd)->common.flags & XNODE_REF_BWIDTH) + 1)
/*
 * It's assumed that the bits we set have been reset.
 * If the existing value is to be changed, the caller is responsible for resetting the original value.
 */
#define XNODE_SET_REF_BWIDTH(cnd, bw)	((cnd)->common.flags |= (bw - 1))

/* Reset the byte-width if already set */
#define XNODE_RESET_REF_BWIDTH(cnd)		((cnd)->common.flags &= ~XNODE_REF_BWIDTH)

typedef enum XMLAddMode
{
	XMLADD_AFTER = 'a',
	XMLADD_BEFORE = 'b',
	XMLADD_INTO = 'i',
	XMLADD_REPLACE = 'r'
} XMLAddMode;

#define XMLADD_MODE(m) (tolower(m))

typedef enum XMLNodeAction
{
	XMLNODE_ACTION_ADD,
	XMLNODE_ACTION_REMOVE
} XMLNodeAction;

/*
 * In-memory structure to represent a single node of the tree.
 */

typedef struct XNodeInternalData
{
	/* Pointer to node or subtree of the storage tree. */
	XMLNodeHdr	node;

	/*
	 * 'true' if 'node' points to a palloc'd copy instead of storage. It
	 * indicates that 'node' has to be freed when the template is no longer
	 * needed.
	 */
	bool		copy;

	XMLNodeContainerData children;
} XNodeInternalData;

typedef struct XNodeInternalData *XNodeInternal;

extern void writeXMLNodeInternal(XNodeInternal node, bool checkElementChildren, char **output, XMLNodeOffset *root);

/*
 * DOM functions
 */
extern Datum xmlnode_children(PG_FUNCTION_ARGS);
extern Datum xmlelement(PG_FUNCTION_ARGS);
extern Datum xmlfragment(PG_FUNCTION_ARGS);

#endif   /* XMLNODE_H */
