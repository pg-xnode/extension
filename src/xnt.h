/*
 * Copyright (C) 2012-2013, Antonin Houska
 */

#ifndef XNT_H_
#define XNT_H_

#include "access/tupdesc.h"
#include "catalog/pg_cast.h"
#include "catalog/pg_proc.h"
#include "catalog/pg_type.h"
#include "utils/array.h"
#include "utils/lsyscache.h"
#include "utils/builtins.h"
#include "utils/syscache.h"
#include "utils/typcache.h"

#include "xpath.h"

#define XNTNODE_NAMESPACE_URI			"http://www.pg-xnode.org/xnt"

#define XNT_TEMPLATE			"template"
#define XNT_TEMPLATE_PRESERVE	0

#define XNT_COPY_OF				"copy-of"
#define XNT_COPY_OF_EXPR		0

#define XNT_ELEMENT				"element"
#define XNT_ELEMENT_NAME		0

#define XNT_ATTRIBUTE			"attribute"
#define XNT_ATTRIBUTE_NAME		0
#define XNT_ATTRIBUTE_VALUE		1


extern XNodeSpecAttributes xntAttributeInfo[];

typedef struct varlena xnttype;
typedef xnttype *xnt;

extern Datum xnode_template_in(PG_FUNCTION_ARGS);
extern Datum xnode_template_out(PG_FUNCTION_ARGS);

extern Datum xnode_from_template(PG_FUNCTION_ARGS);

#endif   /* XNT_H_ */
