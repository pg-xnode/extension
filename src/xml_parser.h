/*
 * Copyright (C) 2012, Antonin Houska
 */

#include "postgres.h"
#include "mb/pg_wchar.h"

#include "xmlnode.h"
#include "xpath.h"

#ifndef XML_PARSER_H
#define XML_PARSER_H

#define XNODE_PARSER_LIST_CHUNK		8

/*
 * This constant is used to estimate how much additional space we need for
 * the binary value, If this value is N, then we allocate strlen(inputText) +
 * 1 / (2^N * strlent(inputText)) for the binary value.
 *
 * This may be implemented as a configuration parameter in future versions.
 */
#define XNODE_OUT_OVERHEAD_BITS		3

/*
 * If changing this,  reflect value of XNODE_OUT_OVERHEAD_BITS.
 *
 * For example it makes little sense to set this to 32 while
 * XNODE_OUT_OVERHEAD_BITS is equal to 3. In such a case, xmlnodeInitParser()
 * function will only add 4 bytes to the input document length when computing
 * output space needed. However, with such small documents the binary size is
 * multiple times bigger than the source text.
 */

#define XNODE_PARSER_OUTPUT_MIN		256

#define XNODE_SPEC_STR_MAX_LEN		16
#define XNODE_DTD_ATTR_TYPES		8
#define XNODE_XDECL_MAX_ATTRS		3
#define XNODE_XDECL_VERSIONS		1

typedef enum XNodeSpecString
{
	XNODE_STR_XDECL_START = 0,
	XNODE_STR_XDECL_END,
	XNODE_STR_DTD_START,
	XNODE_STR_DTD_END,
	XNODE_STR_CMT_START,
	XNODE_STR_CMT_END,
	XNODE_STR_CDATA_START,
	XNODE_STR_CDATA_END,
	XNODE_STR_PI_START,
	XNODE_STR_PI_END
} XNodeSpecString;

typedef enum XNodeSpecStringDTD
{
	XNODE_STR_DTD_ELEMENT = 0,
	XNODE_STR_DTD_ATTLIST,
	XNODE_STR_DTD_ENTITY,
	XNODE_STR_DTD_NOTATION,
	XNODE_STR_DTD_SYSTEM,
	XNODE_STR_DTD_PUBLIC,
	XNODE_STR_DTD_EMPTY,
	XNODE_STR_DTD_ANY,
	XNODE_STR_DTD_PCDATA,
	XNODE_STR_DTD_NDATA,
	XNODE_STR_DTD_REQUIRED,
	XNODE_STR_DTD_IMPLIED,
	XNODE_STR_DTD_FIXED
}	XNodeSpecStringDTD;

typedef enum XNodeXDeclAttNames
{
	XNODE_XDECL_ATNAME_VERSION = 0,
	XNODE_XDECL_ATNAME_ENCODING = 1,
	XNODE_XDECL_ATNAME_STANDALONE = 2
} XNodeXDeclAttNames;

/*
 * Only XNODE_STR_CDATA_END or XNODE_STR_CMT_END expected as 'i'
 */
#define XNODE_SPEC_TEXT_END(i)	((state->srcPos + 2 < state->sizeIn) && strncmp(state->c, specStrings[i],	strlen(specStrings[i])) == 0)


typedef struct XMLParserStateData
{
	/*
	 * Besides document parsing, the code is sometimes used to check attribute
	 * value in a different context (e.g. when 'element constructor receives
	 * the attributes in array)
	 *
	 * 'attrValue' must be set to 'true' in such special cases.
	 */
	bool		attrValue;

	char	   *inputText;
	unsigned int sizeIn;
	unsigned int srcPos,
				dstPos;
	unsigned int srcRow,
				srcRowIncr;
	unsigned int srcCol;

	char	   *c;
	int			cWidth;

	unsigned int nestLevel;

	/*
	 * The parsed document
	 */
	char	   *result,
			   *tree;
	unsigned int sizeOut;
	bool		saveHeader;
	XMLNodeContainerData stack;
	XMLDecl		decl;
} XMLNodeParserStateData;

typedef struct XMLParserStateData *XMLParserState;

#define XNODE_INPUT_END(state)	(*(state)->c == '\0')

#define UNEXPECTED_CHARACTER elog(ERROR, "Unexpected character at row %u, column %u.",\
	state->srcRow, state->srcCol)

extern void initXMLParserState(XMLParserState state, char *inputText, bool attrValue);
extern void finalizeXMLParserState(XMLParserState state);

extern void xmlnodeParseDoc(XMLParserState state);
extern void xmlnodeParseNode(XMLParserState state);
extern char *readXMLAttValue(XMLParserState state, bool output, bool *refs);
extern bool xmlAttrValueIsNumber(char *value);

extern void xmlnodeDumpNode(char *input, XMLNodeOffset nodeOff,
				char **output, unsigned int *pos);
extern char *dumpXMLDecl(XMLDecl decl);

#endif   /* XML_PARSER_H */
