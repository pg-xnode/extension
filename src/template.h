/*
 * Copyright (C) 2012, Antonin Houska
 */

#ifndef TEMPLATE_H_
#define TEMPLATE_H_

#include "xml_parser.h"
#include "xnt.h"

typedef struct XMLTemplateHeaderData
{
	/*
	 * Number substitution nodes (or rather their offsets). These immediately
	 * follow the structure.
	 */
	unsigned short substNodesCount;

	/*
	 * Number of *unique* parameters, i.e. number of parameter names following
	 * the substitution node offsets;
	 */
	unsigned short paramCount;
} XMLTemplateHeaderData;

#define XNODE_ALIGNOF_TEMPL_HDR ALIGNOF_SHORT

typedef struct XMLTemplateHeaderData *XMLTemplateHeader;

typedef struct XMLParamNameSorted
{
	unsigned short order;
	char	   *name;
} XMLParamNameSorted;

/*
 * If attribute value contains xpath expressions, it's understood as a sequence of
 * tokens where each token is either a NULL-terminated string or XPathExpressionData.
 */
#define XMLTEMPL_ATTR_VALUE_MAX_TOKENS		16


extern void parseXMLTemplateNode(XMLParserState state,
	 XMLParserNodeInfo nodeInfo, int specialNodeKind, unsigned int attrCount,
					 XNodeListItem *attrOffsets, unsigned int nmspDecls,
					 unsigned int *attrCountNew, bool **specAttrsValid,
					 unsigned int *specAttrCount);
extern XMLParamNameSorted *getXMLTemplateParamNames(ArrayType *parNameArray,
					   unsigned int templateParamCount, char **templParNames,
						 unsigned short *paramMap);
extern XPathExprOperandValue getXMLTemplateParamValues(Datum row,
	   unsigned int templateParamCount, XPathExprState exprState, Oid fnOid);
extern XPathExpression substituteXMLTemplateParams(XPathExprState exprState,
			   XPathExpression expression, XPathExprOperandValue paramValues,
							unsigned short *paramMap);
extern char *preprocessXMLTemplateAttrValues(XNodeListItem *attrOffsets,
		 unsigned short attrCount, char *parserOutput, unsigned int *outSize,
						   XMLNodeContainer paramNames, bool acceptLocPaths);
extern char *dumpXMLAttrBinaryValue(char *binValue, char **paramNames,
				 XPathExprOperandValue paramValues, unsigned short *paramMap,
					   XPathExprState exprState);
extern XMLNodeHdr getNewXMLAttribute(char *name, uint8 flags, char *value,
				   char decideOnDelim, unsigned int *size);
extern XPathHeader getXPathExpressionForXMLTemplate(char *src,
						 unsigned int termFlags, XMLNodeContainer paramNames,
		unsigned short *endPos, unsigned short *outSize, bool accepLcoPaths);
extern char *getAttrValueForXMLTemplate(char *attrValue, unsigned short *valueSize,
						   XMLNodeContainer paramNames, bool acceptLocPaths);

#endif   /* TEMPLATE_H_ */
