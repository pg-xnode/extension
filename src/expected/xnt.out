set search_path to xml;
CREATE TABLE templates (
	id integer NOT NULL,
	template xnt
);
INSERT INTO templates VALUES 
-- the simplest case: only attribute values are parametrized
(0, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{$p1}" y="{$p2}"/></xnt:template>'),
-- in addition, parameter is used to generate the element name as well as attribute names
(1, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt">
<xnt:element name="point_{$p1}">
	<xnt:attribute name="x_{$p1}" value="{$p2}"/>
	<xnt:attribute name="y_{$p2}" value="{$p3}"/>
</xnt:element>
</xnt:template>'),
-- node can also be used as parameter
(2, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt">
<area><xnt:copy-of expr="$p1"/></area>
</xnt:template>'),
(3, '<xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{1}" y="{2}"/></xnt:template>');
SELECT id, template
FROM templates
ORDER BY id;
 id |                                              template                                               
----+-----------------------------------------------------------------------------------------------------
  0 | <xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{$p1}" y="{$p2}"/></xnt:template>
  1 | <xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt">                                             +
    | <xnt:element name="point_{$p1}">                                                                   +
    |         <xnt:attribute name="x_{$p1}" value="{$p2}"/>                                              +
    |         <xnt:attribute name="y_{$p2}" value="{$p3}"/>                                              +
    | </xnt:element>                                                                                     +
    | </xnt:template>
  2 | <xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt">                                             +
    | <area><xnt:copy-of expr="$p1"/></area>                                                             +
    | </xnt:template>
  3 | <xnt:template xmlns:xnt="http://www.pg-xnode.org/xnt"><point x="{1.00}" y="{2.00}"/></xnt:template>
(4 rows)

CREATE TABLE points (
	id integer NOT NULL,
	x float NOT NULL,
	y float NOT NULL
);
INSERT INTO points VALUES
(0, 0.0, 0.0), (1, 0.3, 5.4), (2, 3.15, 0.01);
CREATE TABLE point_nodes (
	id integer NOT NULL,
	p node NOT NULL
);
INSERT INTO point_nodes
SELECT p.id, xml.node(t.template, '{p1, p2}', ROW (x, y))
FROM templates t, points p
WHERE t.id=0;
SELECT *
FROM point_nodes
ORDER BY id;
 id |             p              
----+----------------------------
  0 | <point x="0" y="0"/>
  1 | <point x="0.3" y="5.4"/>
  2 | <point x="3.15" y="0.01"/>
(3 rows)

SELECT xml.node(template, '{p1, p2, p3}', ROW (p.id, p.x, p.y))
FROM templates t, points p
WHERE t.id=1
ORDER BY p.id;
                node                 
-------------------------------------
 <point_0 x_0="0" y_0="0"/>
 <point_1 x_1="0.3" y_0.3="5.4"/>
 <point_2 x_2="3.15" y_3.15="0.01"/>
(3 rows)

WITH fragments(f) AS (
	-- order the rows so that the fragment is never different from the expected test output.  
	SELECT xml.fragment(p ORDER BY id)
	FROM point_nodes
)
SELECT xml.node(t.template, '{p1}', ROW(f))
FROM fragments f,
templates t
WHERE t.id=2;
                                        node                                         
-------------------------------------------------------------------------------------
 <area><point x="0" y="0"/><point x="0.3" y="5.4"/><point x="3.15" y="0.01"/></area>
(1 row)

DROP TABLE templates;
DROP TABLE points;
DROP TABLE point_nodes;
